File: phpoc-web/src/commonplace/commonplace_chain.js
/**
 * commonplace_chain.js — Commonplace sealed chain (ADR-031).
 *
 * A separate, sealed append-only chain holding its own `commonplace_genesis`
 * followed by `commonplace` day blocks. It shares the same Master Key as the
 * activity ledger (same seed → same MK) but is a structurally independent
 * chain with its own genesis, entry schema (`title`/`tags`/`entry`, optional
 * `ad_hoc`), and history — never mixing with the activity ledger (D7).
 *
 * Web contract (mirrors web LedgerChain + Flutter CommonplaceChain):
 *   new CommonplaceChain(crypto, store, masterKey, identitySecret = null)
 *   - reads/writes StorageBackend under key "commonplace:blocks"
 *   - buildGenesis() appends immediately; buildDayBlock() does NOT append
 *   - the genesis mirrors Flutter's flattened identity shape (NOT the web
 *     ledger's nested `identity: {...}`) for cross-client byte parity.
 *
 * Mirrors `LedgerChain` (Axiom B5) so cross-client parity stays cheap.
 */

import {
  jsonSort,
  computeEntryHash,
  verifyEntryHash,
  computeContentHash,
  isFormatVersionAtLeast,
  CONTENT_HASH_REQUIRED_VERSION,
  ZERO_HASH_64,
} from '../ledger/utils.js';
import { selectSealFields, computeSeal } from '../ledger/seal_fields.js';

const BLOCKS_KEY = 'commonplace:blocks';

export class CommonplaceChain {
  /**
   * @param {object} crypto - CryptoService-like object with seal/verifySeal/mac/verifyMac/encrypt/decrypt/sha256.
   * @param {import('../sync/storage.js').StorageBackend} store - StorageBackend instance.
   * @param {string} masterKey - Hex master key for sealing/encryption.
   * @param {string|null} [identitySecret=null] - Optional identity secret for MAC computation.
   */
  constructor(crypto, store, masterKey, identitySecret = null) {
    this.crypto = crypto;
    this.store = store;
    this.masterKey = masterKey;
    this.identitySecret = identitySecret;
  }

  // ── Store access helpers ──────────────────────────────────────────

  async _getBlocks() {
    const blocks = await this.store.get(BLOCKS_KEY);
    return Array.isArray(blocks) ? blocks : [];
  }

  async _saveBlocks(blocks) {
    await this.store.set(BLOCKS_KEY, blocks);
  }

  // ── Seal / MAC helpers ────────────────────────────────────────────

  /** Verify an HMAC-like seal over a dict (mirrors LedgerChain.verifySeal). */
  verifySeal(data, sealHex) {
    if (this.crypto.verifySeal(jsonSort(data), sealHex, this.masterKey)) {
      return true;
    }
    const canonicalJson = jsonSort(data);
    if (this.crypto.hmacHex(this.masterKey, canonicalJson) === sealHex) {
      return true;
    }
    return false;
  }

  /** Verify an identity MAC over a hash. */
  verifyIdentityMac(dataStr, macTag) {
    return this.crypto.verifyMac(dataStr, macTag, this.identitySecret);
  }

  // ── Block access ──────────────────────────────────────────────────

  async getBlockCount() {
    const blocks = await this._getBlocks();
    return blocks.length;
  }

  async getBlock(index) {
    const blocks = await this._getBlocks();
    if (blocks.length === 0) return null;
    if (index < 0) index = blocks.length + index;
    if (index < 0 || index >= blocks.length) return null;
    return blocks[index];
  }

  async getLastBlock() {
    const blocks = await this._getBlocks();
    return blocks.length > 0 ? blocks[blocks.length - 1] : null;
  }

  async readAll() {
    return await this._getBlocks();
  }

  async getDayBlocks() {
    const blocks = await this._getBlocks();
    return blocks.filter((b) => b.type === 'commonplace');
  }

  /** Resolve the hash key for a Commonplace block (genesis→block_hash, day→day_hash). */
  getBlockHashFor(block) {
    const type = block && block.type;
    if (type === 'commonplace_genesis') return block.block_hash || '';
    if (type === 'commonplace') return block.day_hash || '';
    return block.block_hash || block.day_hash || '';
  }

  /** The cached master-key hex the chain seals under. */
  getMasterKeyHex() {
    return this.masterKey;
  }

  // ── Block building ────────────────────────────────────────────────

  /**
   * Build and APPEND the Commonplace genesis block (mirrors Flutter — the
   * Commonplace genesis appends immediately so a fresh chain is verifiable).
   *
   * @param {object} opts
   * @param {string} opts.username
   * @param {string} opts.email
   * @param {string} opts.recoverySeedEnc
   * @param {string} opts.identityPubKey
   * @param {string} opts.identitySecretEncFallback
   * @param {string} [opts.formatVersion='0.4.0']
   * @returns {Promise<object>} The genesis block.
   */
  async buildGenesis({
    username,
    email,
    recoverySeedEnc,
    identityPubKey,
    identitySecretEncFallback,
    formatVersion = '0.4.0',
  }) {
    if ((await this.getBlockCount()) > 0) {
      throw new Error('Commonplace chain already has blocks — cannot create genesis');
    }

    const gen = {
      type: 'commonplace_genesis',
      day_index: 0,
      prev_hash: ZERO_HASH_64,
      entries: [],
      format_version: formatVersion,
      key_version: 1,
      username,
      email,
      recovery_seed_enc: recoverySeedEnc,
      identity_pub_key: identityPubKey,
      identity_secret_enc_fallback: identitySecretEncFallback,
    };

    // block_hash = seal over the canonical genesis whitelist fields.
    gen.block_hash = computeSeal(gen, this.crypto, this.masterKey);

    if (this.identitySecret) {
      gen.identity_seal = this.crypto.mac(gen.block_hash, this.identitySecret);
    }

    await this._saveBlocks([gen]);
    return gen;
  }

  /**
   * Build a Commonplace day block (does NOT append unless the caller appends).
   *
   * Raw Commonplace entries ({title, tags, entry[, ad_hoc], timestamp_ms[, date]})
   * are encrypted here — `title`/`entry`/`tags`/`ad_hoc` become `_enc` fields so
   * nothing is plaintext at rest (D2) — and content hashes are computed over the
   * encrypted data. Already-encrypted `{hash, data}` maps pass through with
   * content hashes recomputed. The block carries type=commonplace.
   *
   * @param {object[]} entries - Raw Commonplace entry dicts or {hash, data} pairs.
   * @param {string} prevHash - Hash of the preceding block.
   * @param {string} dateStr - ISO date string (YYYY-MM-DD).
   * @param {number} [keyVersion=1] - Key version (default 1).
   * @returns {Promise<object>} The constructed day block.
   */
  async buildDayBlock(entries, prevHash, dateStr, keyVersion = 1) {
    const existingDays = (await this.getDayBlocks()).length;
    const dayIndex = existingDays === 0 ? 1 : existingDays + 1;

    const normalizedEntries = [];
    for (const entry of entries) {
      let data;
      let alreadySealed = false;
      if (entry && typeof entry === 'object' && 'data' in entry) {
        data = Object.assign({}, entry.data);
        alreadySealed = true;
      } else {
        data = Object.assign({}, entry);
      }

      // Strip staging-only fields.
      delete data.is_active;
      delete data.unsealed;
      delete data.entry_id;
      delete data.device_uuid;
      delete data.hash;

      if (!alreadySealed) {
        data = this._encryptCommonplaceEntry(data, dateStr);
      }

      // Compute & record the content hash over the (encrypted) entry data.
      data.content_hash = computeContentHash(data, this.crypto, this.masterKey);

      normalizedEntries.push({
        hash: computeEntryHash(data, this.crypto),
        data,
      });
    }

    const block = {
      type: 'commonplace',
      date: dateStr,
      day_index: dayIndex,
      prev_hash: prevHash,
      entries: normalizedEntries,
      key_version: keyVersion,
    };

    block.day_hash = computeSeal(block, this.crypto, this.masterKey);

    if (this.identitySecret) {
      block.identity_seal = this.crypto.mac(block.day_hash, this.identitySecret);
    }

    return block;
  }

  // ── Append / truncate ─────────────────────────────────────────────

  /** Reject block types that don't belong to the Commonplace chain (ADR-029a). */
  _assertAllowedType(block) {
    const type = block && block.type;
    if (type !== 'commonplace' && type !== 'commonplace_genesis') {
      throw new Error(`Unknown/foreign block type for Commonplace chain: ${type}`);
    }
  }

  async append(block) {
    this._assertAllowedType(block);
    const blocks = await this._getBlocks();
    if (blocks.length > 0) {
      const last = blocks[blocks.length - 1];
      const expectedPrev = this.getBlockHashFor(last);
      const actualPrev = block.prev_hash || '';
      if (expectedPrev && actualPrev !== expectedPrev) {
        throw new Error(
          `prev_hash mismatch: expected ${expectedPrev.slice(0, 12)}…, got ${actualPrev.slice(0, 12)}…`
        );
      }
    }
    blocks.push(block);
    await this._saveBlocks(blocks);
  }

  async appendBlocks(blocks) {
    if (!blocks || blocks.length === 0) return;
    for (const b of blocks) this._assertAllowedType(b);

    const existing = await this._getBlocks();

    if (existing.length > 0) {
      const expectedPrev = this.getBlockHashFor(existing[existing.length - 1]);
      const firstPrev = blocks[0].prev_hash || '';
      if (expectedPrev && firstPrev !== expectedPrev) {
        throw new Error(
          `Bridge prev_hash mismatch: expected ${expectedPrev.slice(0, 12)}…, got ${firstPrev.slice(0, 12)}…`
        );
      }
    }

    for (let i = 1; i < blocks.length; i++) {
      const expected = this.getBlockHashFor(blocks[i - 1]);
      const actual = blocks[i].prev_hash || '';
      if (expected && actual !== expected) {
        throw new Error(
          `Internal prev_hash mismatch at index ${i}: expected ${expected.slice(0, 12)}…, got ${actual.slice(0, 12)}…`
        );
      }
    }

    existing.push(...blocks);
    await this._saveBlocks(existing);
  }

  /**
   * Remove `removeCount` blocks from the end, preserving at least the genesis.
   * @returns {Promise<object[]>} The removed blocks.
   */
  async truncate(removeCount) {
    const blocks = await this._getBlocks();
    if (removeCount <= 0 || blocks.length === 0) return [];
    const keepCount = Math.max(1, blocks.length - removeCount);
    if (keepCount >= blocks.length) return [];
    const removed = blocks.splice(keepCount);
    await this._saveBlocks(blocks);
    return removed;
  }

  // ── Verification ──────────────────────────────────────────────────

  async verify() {
    const blocks = await this._getBlocks();
    if (blocks.length === 0) return true;

    const genesis = blocks[0];
    const genesisKv = genesis.key_version ?? 1;
    const requireContentHash = isFormatVersionAtLeast(genesis, CONTENT_HASH_REQUIRED_VERSION);

    for (let i = 0; i < blocks.length; i++) {
      const block = blocks[i];

      // prev_hash linkage
      if (i > 0 && block.prev_hash !== this.getBlockHashFor(blocks[i - 1])) {
        return false;
      }

      // Block seal
      if (!this._verifyBlockSeal(block)) return false;

      // Identity seal (optional signing parity)
      if (this.identitySecret && 'identity_seal' in block) {
        const hash = this.getBlockHashFor(block);
        if (!this.crypto.verifyMac(hash, block.identity_seal, this.identitySecret)) {
          return false;
        }
      }

      // Entry hashes + content hashes + key-version invariant (day blocks)
      if (block.type === 'commonplace') {
        for (const entry of block.entries || []) {
          const data = entry && entry.data;
          const hash = entry && entry.hash;
          if (!data || !hash) return false;

          if (!verifyEntryHash(data, hash, this.crypto)) return false;

          const contentHash = data.content_hash;
          const hasContentHash = contentHash !== undefined && contentHash !== null && contentHash !== '';
          if (requireContentHash && !hasContentHash) return false;
          if (hasContentHash && computeContentHash(data, this.crypto, this.masterKey) !== data.content_hash) return false;
        }

        const blockKv = block.key_version ?? 1;
        if (blockKv > genesisKv) return false;
      }
    }

    return true;
  }

  /** Verify a single block's seal against the ADR-029a per-type whitelist. */
  _verifyBlockSeal(block) {
    const type = block && block.type;
    let hashKey;
    if (type === 'commonplace_genesis') hashKey = 'block_hash';
    else if (type === 'commonplace') hashKey = 'day_hash';
    else return false;

    let checkData;
    try {
      checkData = selectSealFields(block);
    } catch (_) {
      return false;
    }

    return this.verifySeal(checkData, block[hashKey]);
  }

  // ── Entry encrypt/decrypt helpers ─────────────────────────────────

  /**
   * Encrypt a raw Commonplace entry's content fields for sealed storage.
   * `title`→`title_enc`, `entry`→`entry_enc`, `tags`→`tags_enc`,
   * `ad_hoc`→`ad_hoc_enc` (only when present & non-empty). `type`,
   * `timestamp_ms`, and `date` stay plaintext (not content). No `comment`
   * field exists in the Commonplace schema (ADR-031 — `entry` replaces it).
   */
  _encryptCommonplaceEntry(data, dateStr) {
    const type = data.type || 'commonplace';
    const title = data.title || '';
    const entryText = data.entry || '';
    const tags = (data.tags || []).map((t) => String(t));
    const adHoc =
      data.ad_hoc && typeof data.ad_hoc === 'object' && !Array.isArray(data.ad_hoc)
        ? Object.assign({}, data.ad_hoc)
        : null;
    const timestampMs = data.timestamp_ms;

    const out = {
      type,
      timestamp_ms: Number.isInteger(timestampMs) ? timestampMs : 0,
      date: data.date || dateStr,
      title_enc: this.crypto.encrypt(title, this.masterKey),
      entry_enc: this.crypto.encrypt(entryText, this.masterKey),
      tags_enc: this.crypto.encrypt(JSON.stringify(tags), this.masterKey),
    };

    if (adHoc && Object.keys(adHoc).length > 0) {
      out.ad_hoc_enc = this.crypto.encrypt(JSON.stringify(adHoc), this.masterKey);
    }

    return out;
  }

  /**
   * Decrypt one Commonplace entry's encapsulated fields back to plaintext.
   * Returns the public read shape: {title, entry, tags, timestamp_ms, date,
   * [ad_hoc], [type]}.
   */
  decryptEntryData(data) {
    const result = {};

    result.title = data.title_enc !== undefined ? this._tryDecrypt(data.title_enc) : '';
    result.entry = data.entry_enc !== undefined ? this._tryDecrypt(data.entry_enc) : '';

    let tags = [];
    if (data.tags_enc !== undefined) {
      try {
        tags = JSON.parse(this.crypto.decrypt(data.tags_enc, this.masterKey)).map((e) => String(e));
      } catch (_) {
        tags = [];
      }
    }
    result.tags = tags;

    if (data.ad_hoc_enc !== undefined) {
      try {
        result.ad_hoc = JSON.parse(this.crypto.decrypt(data.ad_hoc_enc, this.masterKey));
      } catch (_) {
        result.ad_hoc = {};
      }
    }

    result.timestamp_ms = data.timestamp_ms;
    result.date = data.date;
    if (data.type !== undefined) result.type = data.type;

    return result;
  }

  _tryDecrypt(ciphertext) {
    try {
      return this.crypto.decrypt(ciphertext, this.masterKey);
    } catch (_) {
      return '';
    }
  }
}

File: phpoc-flutter/lib/data/commonplace/commonplace_chain.dart
import 'package:phpoc_flutter/core/crypto/crypto_service.dart';
import 'package:phpoc_flutter/data/ledger/helpers.dart'
    show
        computeEntryHash,
        computeContentHash,
        verifyEntryHashTwoWay,
        verifyContentHash,
        compareVersions,
        secretToHex;
import 'package:phpoc_flutter/data/ledger/sealable_chain.dart';
import 'package:phpoc_flutter/data/ledger/chain_reconcile.dart';
import 'dart:convert';

/// The Commonplace chain — a separate, sealed append-only chain (ADR-031).
///
/// Holds its own `commonplace.json` sequence: a `commonplace_genesis` block
/// followed by `commonplace` day blocks. It shares the same Master Key as the
/// activity ledger (same seed → same MK) but is a structurally independent
/// chain with its own genesis, entry schema (`title`/`tags`/`entry`, optional
/// `ad_hoc`), and history — never mixing with the activity ledger (D7).
///
/// Mirrors `LedgerChain` (Axiom B5) so cross-client parity stays cheap.
class CommonplaceChain with SealableChain {
  @override
  final CryptoService crypto;
  final dynamic store;
  final String? identitySecret;
  final String? _identitySecretHex;

  CommonplaceChain({
    required this.crypto,
    required this.store,
    this.identitySecret,
  }) : _identitySecretHex = secretToHex(identitySecret);

  /// Hex identity secret for SealableChain / identity-signing parity.
  @override
  String? get identitySecretHex => _identitySecretHex;

  /// Canonical per-type seal-field whitelist (ADR-029a) for Commonplace blocks.
  @override
  final Map<String, List<String>> sealFieldsByType = {
    'commonplace_genesis': ['type', 'day_index', 'date', 'prev_hash', 'entries', 'original_hash'],
    'commonplace': ['type', 'day_index', 'date', 'prev_hash', 'entries', 'original_hash'],
  };

  /// The value key that holds a Commonplace block's own seal.
  @override
  String? hashKeyFor(String? type) {
    switch (type) {
      case 'commonplace_genesis':
        return 'block_hash';
      case 'commonplace':
        return 'day_hash';
      default:
        return null;
    }
  }


  // ═══════════════════════════════════════════════════════════════
  // Block Building
  // ═══════════════════════════════════════════════════════════════

  /// Build and APPEND the Commonplace genesis block.
  ///
  /// Unlike `LedgerChain.buildGenesisBlock` (which only returns the block),
  /// the Commonplace genesis is appended immediately so a freshly created
  /// chain is immediately verifiable and day blocks can link onto it.
  ///
  /// Throws if the chain already has blocks.
  Map<String, dynamic> buildGenesis({
    required String username,
    required String email,
    required String recoverySeedEnc,
    required String identityPubKey,
    required String identitySecretEncFallback,
    String formatVersion = '0.4.0',
  }) {
    if (getBlockCount() > 0) {
      throw Exception(
          'Commonplace chain already has blocks — cannot create genesis');
    }

    final gen = <String, dynamic>{
      'type': 'commonplace_genesis',
      'day_index': 0,
      'prev_hash': '0' * 64,
      'entries': <Map<String, dynamic>>[],
      'format_version': formatVersion,
      'key_version': 1,
      'username': username,
      'email': email,
      'recovery_seed_enc': recoverySeedEnc,
      'identity_pub_key': identityPubKey,
      'identity_secret_enc_fallback': identitySecretEncFallback,
    };

    // Compute block_hash: seal over the canonical genesis fields.
    final blockHash = sealBlock(gen);
    gen['block_hash'] = blockHash;

    // Identity seal over block_hash (optional signing parity with ledger).
    if (_identitySecretHex != null) {
      gen['identity_seal'] = crypto.sign(blockHash, _identitySecretHex);
    }

    store.appendBlocks([gen]);
    return gen;
  }

  /// Build a Commonplace day block (does NOT append unless caller appends).
  ///
  /// [entries] may contain pre-hashed `{hash, data}` maps or raw dicts. Raw
  /// Commonplace entries (schema `{title, tags, entry[, ad_hoc], timestamp_ms}`)
  /// are encrypted here — `title`/`entry`/`tags`/`ad_hoc` become `_enc` fields
  /// so nothing is plaintext at rest (D2) — and their content hashes are
  /// computed over the encrypted data. Already-encrypted `{hash, data}` maps
  /// pass through with content hashes recomputed. The block carries
  /// type=commonplace.
  Map<String, dynamic> buildDayBlock({
    required List<dynamic> entries,
    required String prevHash,
    required String dateStr,
    int? keyVersion,
  }) {
    final existingDays = countDayBlocks();
    final dayIndex = existingDays == 0 ? 1 : existingDays + 1;
    final kv = keyVersion ?? 1;

    // Normalize entries: encrypt raw fields, strip staging-only fields, and
    // always recompute content + entry hashes from data.
    final normalizedEntries = <Map<String, dynamic>>[];
    for (final entry in entries) {
      Map<String, dynamic> data;
      bool alreadySealed = false;
      if (entry is Map && entry.containsKey('data')) {
        data = Map<String, dynamic>.from(entry['data'] as Map);
        alreadySealed = true;
      } else {
        data = Map<String, dynamic>.from(entry as Map);
      }

      // Strip staging-only fields.
      data.remove('is_active');
      data.remove('unsealed');
      data.remove('entry_id');
      data.remove('device_uuid');
      data.remove('hash');

      if (!alreadySealed) {
        data = _encryptCommonplaceEntry(data, dateStr: dateStr);
      }

      // Compute & record the content hash over the encrypted entry data.
      data['content_hash'] = computeContentHash(data, crypto);

      final computedHash = computeEntryHash(data);
      normalizedEntries.add({
        'hash': computedHash,
        'data': data,
      });
    }

    final block = <String, dynamic>{
      'type': 'commonplace',
      'date': dateStr,
      'day_index': dayIndex,
      'prev_hash': prevHash,
      'entries': normalizedEntries,
      'key_version': kv,
    };

    final dayHash = sealBlock(block);
    block['day_hash'] = dayHash;

    if (_identitySecretHex != null) {
      block['identity_seal'] = crypto.sign(dayHash, _identitySecretHex);
    }

    return block;
  }

  // ═══════════════════════════════════════════════════════════════
  // Mutations
  // ═══════════════════════════════════════════════════════════════

  /// Append a single Commonplace block, verifying prev_hash linkage.
  ///
  /// Rejects blocks whose type is not `commonplace` or `commonplace_genesis`
  /// (the activity logger's block types must not leak into this chain,
  /// ADR-029a per-type whitelist).
  void append(Map<String, dynamic> block) {
    _assertAllowedType(block);
    final last = getLastBlock();
    if (last != null) {
      final expectedPrev = getBlockHashFor(last);
      final actualPrev = block['prev_hash'] as String? ?? '';
      if (expectedPrev.isNotEmpty && actualPrev != expectedPrev) {
        throw Exception(
          'prev_hash mismatch: expected $expectedPrev, got $actualPrev',
        );
      }
    }
    store.appendBlocks([block]);
  }

  /// Append multiple Commonplace blocks with full linkage verification.
  void appendBlocks(List<Map<String, dynamic>> blocks) {
    if (blocks.isEmpty) return;
    for (final b in blocks) {
      _assertAllowedType(b);
    }

    final last = getLastBlock();

    // Bridge linkage.
    if (last != null) {
      final expectedPrev = getBlockHashFor(last);
      final firstPrev = blocks.first['prev_hash'] as String? ?? '';
      if (expectedPrev.isNotEmpty && firstPrev != expectedPrev) {
        throw Exception(
          'Bridge prev_hash mismatch: expected $expectedPrev, got $firstPrev',
        );
      }
    }

    // Internal linkage.
    for (var i = 1; i < blocks.length; i++) {
      final expected = getBlockHashFor(blocks[i - 1]);
      final actual = blocks[i]['prev_hash'] as String? ?? '';
      if (expected.isNotEmpty && actual != expected) {
        throw Exception(
          'Internal prev_hash mismatch at index $i: '
          'expected $expected, got $actual',
        );
      }
    }

    store.appendBlocks(blocks);
  }

  /// Remove [removeCount] blocks from the end of the chain.
  List<Map<String, dynamic>> truncate(int removeCount) {
    final currentCount = getBlockCount();
    if (currentCount == 0) return [];

    int keepCount = currentCount - removeCount;
    if (keepCount < 0) keepCount = 0;
    if (keepCount == 0 && removeCount > currentCount) keepCount = 1;

    return store.truncate(keepCount);
  }

  /// Append-only merge of [remoteBlocks] onto this chain.
  ///
  /// Mirrors `SyncService.reconcileRemoteLedger` semantics for the Commonplace
  /// sealed chain (ADR-031 remote-sync slice):
  ///   - a remote block identical (same hash) to the local one is skipped;
  ///   - a remote tail that bridges the last local block is appended in order;
  ///   - same index / different hash, a non-bridging tip, or a non-genesis
  ///     first block on an empty chain is reported as a conflict and **never
  ///     written** (a stale device never clobbers the remote canonical chain).
  CommonplaceReconcileResult reconcileRemoteChain(
    List<Map<String, dynamic>> remoteBlocks,
  ) {
    final r = reconcileChainCore(
      local: readAll(),
      remoteBlocks: remoteBlocks,
      blockHash: getBlockHashFor,
      genesisType: 'commonplace_genesis',
      appendBlocks: appendBlocks,
    );
    return CommonplaceReconcileResult(
      conflictedIndices: r.conflictedIndices,
      appended: r.appended,
    );
  }

  // ═══════════════════════════════════════════════════════════════
  // Verification
  // ═══════════════════════════════════════════════════════════════

  /// Verify the entire Commonplace chain.
  bool verify() => verifyBlocks(readAll());

  /// Verify an arbitrary Commonplace block sequence (used by both [verify] and
  /// the pull-service pre-import validation). An empty list is vacuously valid.
  bool verifyBlocks(List<Map<String, dynamic>> blocks) {
    if (blocks.isEmpty) return true;

    final genesis = blocks.first;
    final genesisKv = genesis['key_version'] as int? ?? 1;
    final formatVersion = genesis['format_version'] as String? ?? '0.4.0';
    final requireContentHash = compareVersions(formatVersion, '0.4.0') >= 0;

    for (var i = 0; i < blocks.length; i++) {
      final block = blocks[i];

      if (!prevHashValid(i > 0 ? blocks[i - 1] : null, block)) {
        return false;
      }

      if (!verifyBlockSeal(block)) return false;

      if (_identitySecretHex != null && block.containsKey('identity_seal')) {
        final hash = getBlockHashFor(block);
        if (!crypto.verifySignature(
            hash, block['identity_seal'] as String, _identitySecretHex)) {
          return false;
        }
      }

      if (block['type'] == 'commonplace') {
        final entries = block['entries'] as List<dynamic>? ?? [];
        for (final entry in entries) {
          if (entry is! Map) return false;
          final data = entry['data'] as Map<String, dynamic>?;
          final hash = entry['hash'] as String?;
          if (data == null || hash == null) return false;

          if (!verifyEntryHashTwoWay(data, hash)) return false;

          final contentHash = data['content_hash'] as String?;
          final hasContentHash = contentHash != null && contentHash.isNotEmpty;
          if (requireContentHash && !hasContentHash) return false;
          if (hasContentHash) {
            if (!verifyContentHash(
              data,
              contentHash,
              decryptFn: (c) => crypto.decryptWithCachedKey(c),
            )) {
              return false;
            }
          }
        }

        // Key-version invariant: a day block may not rotate past its genesis.
        final blockKv = block['key_version'] as int? ?? 1;
        if (blockKv > genesisKv) return false;
      }
    }

    return true;
  }

  // ═══════════════════════════════════════════════════════════════
  // ═══════════════════════════════════════════════════════════════
  // Accessors
  // ═══════════════════════════════════════════════════════════════

  /// Resolve the hashing key for a Commonplace block (SealableChain).
  ///
  /// `commonplace_genesis` → `block_hash`; `commonplace` → `day_hash`;
  /// anything else → empty string (mirrors ledger `getBlockHash`).
  @override
  String getBlockHashFor(Map<String, dynamic> block) {
    final type = block['type'] as String?;
    switch (type) {
      case 'commonplace_genesis':
        return block['block_hash'] as String? ?? '';
      case 'commonplace':
        return block['day_hash'] as String? ?? '';
      default:
        return '';
    }
  }

  /// The cached master-key hex the chain seals under.
  String? getMasterKeyHex() => crypto.getMasterKey();

  /// All blocks in the chain.
  List<Map<String, dynamic>> readAll() {
    return store.readBlocks();
  }

  /// Total number of blocks.
  int getBlockCount() {
    return store.getBlockCount();
  }

  /// Only Commonplace day-type blocks (excludes genesis).
  @override
  List<Map<String, dynamic>> getDayBlocks() {
    return readAll().where((b) => b['type'] == 'commonplace').toList();
  }

  /// The last block, or null if the chain is empty.
  Map<String, dynamic>? getLastBlock() {
    return store.getLastBlock();
  }

  // ═══════════════════════════════════════════════════════════════
  // Internal helpers
  // ═══════════════════════════════════════════════════════════════

  /// Encrypt a raw Commonplace entry's content fields for sealed storage.
  ///
  /// `title`→`title_enc`, `entry`→`entry_enc`, `tags`→`tags_enc`,
  /// `ad_hoc`→`ad_hoc_enc` (only when present). `type`, `timestamp_ms`, and
  /// `date` stay plaintext (they are not content). There is no `comment` field
  /// in the Commonplace schema (ADR-031 — `entry` replaces it).
  Map<String, dynamic> _encryptCommonplaceEntry(
      Map<String, dynamic> data, {
      required String dateStr}) {
    final type = data['type'] as String? ?? 'commonplace';
    final title = data['title'] as String? ?? '';
    final entryText = data['entry'] as String? ?? '';
    final tags = (data['tags'] as List<dynamic>? ?? <dynamic>[])
        .map((t) => t.toString())
        .toList();
    final adHoc = data.containsKey('ad_hoc') && data['ad_hoc'] is Map
        ? Map<String, dynamic>.from(data['ad_hoc'] as Map)
        : null;
    final timestampMs = data['timestamp_ms'];

    return <String, dynamic>{
      'type': type,
      'timestamp_ms': timestampMs is int ? timestampMs : 0,
      'date': data['date'] as String? ?? dateStr,
      'title_enc': crypto.encryptWithCachedKey(title),
      'entry_enc': crypto.encryptWithCachedKey(entryText),
      'tags_enc': crypto.encryptWithCachedKey(jsonEncode(tags)),
      if (adHoc != null && adHoc.isNotEmpty)
        'ad_hoc_enc': crypto.encryptWithCachedKey(jsonEncode(adHoc)),
    };
  }

  /// Throw unless [block]'s type belongs to the Commonplace whitelist.
  void _assertAllowedType(Map<String, dynamic> block) {
    final type = block['type'] as String?;
    if (type != 'commonplace' && type != 'commonplace_genesis') {
      throw Exception('Unknown/foreign block type for Commonplace chain: $type');
    }
  }

  /// Decrypt one Commonplace entry's encapsulated fields back to plaintext.
  ///
  /// Returns a map with `title`, `entry`, `tags`, `timestamp_ms`, `date`,
  /// optional `ad_hoc`, and `type` (the public read shape).
  Map<String, dynamic> decryptEntryData(Map<String, dynamic> data) {
    final result = <String, dynamic>{};

    var title = '';
    if (data.containsKey('title_enc')) {
      try {
        title = crypto.decryptWithCachedKey(data['title_enc'] as String);
      } catch (_) {}
    }
    result['title'] = title;

    var entry = '';
    if (data.containsKey('entry_enc')) {
      try {
        entry = crypto.decryptWithCachedKey(data['entry_enc'] as String);
      } catch (_) {}
    }
    result['entry'] = entry;

    var tags = <String>[];
    if (data.containsKey('tags_enc')) {
      try {
        final plain = crypto.decryptWithCachedKey(data['tags_enc'] as String);
        tags = (jsonDecode(plain) as List<dynamic>).map((e) => e.toString()).toList();
      } catch (_) {}
    }
    result['tags'] = tags;

    if (data.containsKey('ad_hoc_enc')) {
      try {
        final plain = crypto.decryptWithCachedKey(data['ad_hoc_enc'] as String);
        result['ad_hoc'] = jsonDecode(plain) as Map<String, dynamic>;
      } catch (_) {
        result['ad_hoc'] = <String, dynamic>{};
      }
    }

    result['timestamp_ms'] = data['timestamp_ms'];
    result['date'] = data['date'];
    if (data.containsKey('type')) {
      result['type'] = data['type'];
    }
    return result;
  }
}

/// Result of a [CommonplaceChain.reconcileRemoteChain] merge: which remote
/// block ordinals diverged from the local sealed chain (never written), and
/// how many missing blocks were appended (behind-device catch-up).
class CommonplaceReconcileResult {
  /// Block ordinals where the remote chain conflicted with the local chain and
  /// was NOT written (fork / same-index-different-hash / non-bridging tip /
  /// non-genesis-first on an empty chain).
  final List<int> conflictedIndices;

  /// Number of missing remote blocks appended to the local chain.
  final int appended;

  const CommonplaceReconcileResult({
    this.conflictedIndices = const [],
    this.appended = 0,
  });

  /// Whether the merge surfaced any divergent/cannot-merge remote block.
  bool get hasConflicts => conflictedIndices.isNotEmpty;
}
